WalletWasabi/Extensions/RPCClientExtensions.cs

```diff
@@ -59,89 +64,71 @@ private static EstimateSmartFeeResponse SimulateRegTestFeeEstimation(int confirm
-			var allFeeEstimate = new AllFeeEstimate(estimateMode, estimations, rpcStatus.Synchronized);
-			return allFeeEstimate;
-		}
 
-		public static async Task<RpcStatus> GetRpcStatusAsync(this IRPCClient rpc, CancellationToken cancel)
-		{
-			try
-			{
-				var bci = await rpc.GetBlockchainInfoAsync().ConfigureAwait(false);
-				cancel.ThrowIfCancellationRequested();
-				var pi = await rpc.GetPeersInfoAsync().ConfigureAwait(false);
+			var estmimations = (simulateIfRegTest && rpc.Network == Network.RegTest)
+				? SimulateRegTestFeeEstimation(estimateMode)
+				: await GetFeeEstimationsAsync(rpc, estimateMode);
 
-				return RpcStatus.Responsive(bci.Headers, bci.Blocks, pi.Length);
-			}
-			catch (Exception ex) when (ex is not OperationCanceledException and not TimeoutException)
-			{
-				Logger.LogTrace(ex);
-				return RpcStatus.Unresponsive;
-			}
+			return new AllFeeEstimate(
+				estimateMode,
+				estmimations,
+				rpcStatus.Synchronized);
 		}
 
-		private static async Task<Dictionary<int, int>> EstimateHalfFeesAsync(this IRPCClient rpc, IDictionary<int, int> estimations, int smallTarget, int smallTargetFee, int largeTarget, int largeTargetFee, FeeRate sanityFeeRate, EstimateSmartFeeMode estimateMode = EstimateSmartFeeMode.Conservative, bool simulateIfRegTest = false)
+		private static async Task<Dictionary<int, int>> GetFeeEstimationsAsync(IRPCClient rpc, EstimateSmartFeeMode estimateMode)
 		{
-			var newEstimations = new Dictionary<int, int>();
-			foreach (var est in estimations)
-			{
-				newEstimations.TryAdd(est.Key, est.Value);
-			}
+			var batchClient = rpc.PrepareBatch();
 
-			if (Math.Abs(smallTarget - largeTarget) <= 1)
-			{
-				return newEstimations;
-			}
+			var rpcFeeEstimationTasks = Constants.ConfirmationTargets
+				.Select(target => batchClient.EstimateSmartFeeAsync(target, estimateMode))
+				.ToList();
+
+			await batchClient.SendBatchAsync().ConfigureAwait(false);
+
+			var allTask = Task.WhenAll(rpcFeeEstimationTasks);
 
-			if (smallTargetFee == 0)
+			try
 			{
-				var smallTargetFeeResult = await rpc.EstimateSmartFeeAsync(smallTarget, sanityFeeRate, estimateMode, simulateIfRegTest).ConfigureAwait(false);
-				smallTargetFee = (int)Math.Ceiling(smallTargetFeeResult.FeeRate.SatoshiPerByte);
-				newEstimations.TryAdd(smallTarget, smallTargetFee);
+				await allTask;
 			}
-
-			if (largeTargetFee == 0)
+			catch
 			{
-				var largeTargetFeeResult = await rpc.EstimateSmartFeeAsync(largeTarget, sanityFeeRate, estimateMode, simulateIfRegTest).ConfigureAwait(false);
-				largeTargetFee = (int)Math.Ceiling(largeTargetFeeResult.FeeRate.SatoshiPerByte);
-
-				// Blocks should never be larger than the target that we asked for, so it's just a sanity check.
-				largeTarget = Math.Min(largeTarget, largeTargetFeeResult.Blocks);
-				newEstimations.TryAdd(largeTarget, largeTargetFee);
+				if (rpcFeeEstimationTasks.All(x => x.IsFaulted))
+				{
+					throw rpcFeeEstimationTasks[0].Exception?.InnerExceptions[0] 
```

Comment:
So I understand that we are in the situation when all `EstimateSmartFeeAsync` failed with unhandled exceptions. So you take first estimate task and retrieve its exception (i.e. `rpcFeeEstimationTasks[0].Exception?`) but why is `?` necessary? `!` would make more sense to me.

Right, and this object is presumably `AggregateException`, so you take first inner. But then why is it necessary to have `?? new Exception($"{nameof(GetFeeEstimationsAsync)} failed to fetch fee estimations.")`?

Reply:
I would never use `! ` because even in the case that you are 100% sure the property is never `null` that's something that can change so, I do not assume it is not `null`, i prefer to make sure the code will behaves well even in those rare cases. Anyway, I am sure there is a better way to write this code but i am also sure using `!` is not the way.

Reply:
I would say that it's subjective. For me, in general, it's the other way around because `?` hides bugs that an IDE may warn about and when everything is with `?` it's impossible to tell what is "defensive approach" and what is by design.